syntax = "proto2";

import "store/common/common-proto.proto";

package strongstore.proto;

message RequestID {
    required uint64 client_id = 1;
    required uint64 client_req_id = 2;
}

message Get {
    required RequestID rid = 1;
    required uint64 transaction_id = 2;
    required string key = 3;
    optional TimestampMessage timestamp = 4;
}

message GetReply {
    required RequestID rid = 1;
    required string key = 2;
    required string val = 3;
}

message RWCommitCoordinator {
    required RequestID rid = 1;
    required uint64 transaction_id = 2;
    required TransactionMessage transaction = 3;
    required int32 n_participants = 4;
}

message RWCommitCoordinatorReply {
    required RequestID rid = 1;
    required int32 status = 2;
    required uint64 commit_timestamp = 3;
}

message RWCommitParticipant {
    required RequestID rid = 1;
    required uint64 transaction_id = 2;
    required TransactionMessage transaction = 3;
    required int32 coordinator_shard = 4;
}

message RWCommitParticipantReply {
    required RequestID rid = 1;
    required int32 status = 2;
}

message PrepareOK {
    required RequestID rid = 1;
    required uint64 transaction_id = 2;
    required int32 participant_shard = 3;
    required uint64 prepare_timestamp = 4;
}

message PrepareOKReply {
    required RequestID rid = 1;
    required int32 status = 2;
    required uint64 commit_timestamp = 3;
}

message GetMessage {
    required string key = 1;
    optional TimestampMessage timestamp = 2;
}

message PrepareMessage {
    required TransactionMessage txn = 1;
    optional uint64 timestamp = 2;
}

message PrepareOKMessage {
    required uint64 timestamp = 1;
    required int32 participantShard = 2;
}

message PrepareAbortMessage {
    required int32 participantShard = 1;
}

message CommitMessage {
    required uint64 timestamp = 1;
    optional TransactionMessage transaction = 2;
}

message AbortMessage {
    required int32 coordinatorShard = 1;
}

message Request {
     enum Operation {
          GET = 1;
          PREPARE = 2;
          PREPARE_OK = 3;
          PREPARE_ABORT = 4;
          COMMIT = 5;
          ABORT = 6;
     }	
     required Operation op = 1;
     required uint64 txnid = 2;
     optional GetMessage get = 3;
     optional PrepareMessage prepare = 4;
     optional PrepareOKMessage prepareOK = 5;
     optional PrepareAbortMessage prepareAbort = 6;
     optional CommitMessage commit = 7;
     optional AbortMessage abort = 8;
}

message Reply {
     // 0 = OK
     // -1 = failed
     // -2 = retry
     // -3 = abstain/no reply
     required int32 status = 1;
     optional string value = 2;
     optional uint64 timestamp = 3;
}
